sound/soc/sof/amd/pci-rn.c

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license. When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2021 Advanced Micro Devices, Inc. All rights reserved.
+//
+// Authors: Ajit Kumar Pandey <AjitKumar.Pandey@amd.com>
+
+/*
+ * PCI interface for Renoir ACP device
+ */
+
+#include <linux/module.h>
+#include <linux/pci.h>
+#include <linux/platform_device.h>
+#include <sound/sof.h>
+#include <sound/soc-acpi.h>
+#include "../ops.h"
+#include "../sof-pci-dev.h"
+#include "../../amd/mach-config.h"
+#include "acp.h"
+
+#define ACP3x_REG_START		0x1240000
+#define ACP3x_REG_END		0x125C000
+
+static struct platform_device *dmic_dev;
+static struct platform_device *pdev;
+
+static const struct resource renoir_res[] = {
+	{
+		.start = 0,
+		.end = ACP3x_REG_END - ACP3x_REG_START,
+		.name = "acp_mem",
+		.flags = IORESOURCE_MEM,
+	},
+	{
+		.start = 0,
+		.end = 0,
+		.name = "acp_dai_irq",
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+static const struct sof_dev_desc renoir_desc = {
+	.machines		= snd_soc_acpi_amd_sof_machines,
+	.resindex_lpe_base	= 0,
+	.resindex_pcicfg_base	= -1,
+	.resindex_imr_base	= -1,
+	.irqindex_host_ipc	= -1,
+	.default_fw_path	= "amd/sof",
+	.default_tplg_path	= "amd/sof-tplg",
+	.default_fw_filename	= "sof-rn.ri",
+	.nocodec_tplg_filename	= "sof-acp.tplg",
+	.ops			= &sof_renoir_ops,
+};
+
+static int acp_pci_rn_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
+{
+	struct device *dev = &pci->dev;
+	struct resource *res;
+	const struct resource *res_i2s = renoir_res;
+	struct platform_device_info pdevinfo;
+	unsigned int flag, i, addr;
+	int num_res, ret;
+
+	num_res = ARRAY_SIZE(renoir_res);
+
+	flag = snd_amd_acp_find_config(pci);
+	if (flag != FLAG_SOF && flag != FLAG_SOF_ONLY_DMIC)
+		return -ENODEV;
+
+	ret = sof_pci_probe(pci, pci_id);
+	if (ret != 0)
+		return -EINVAL;
+
+	dmic_dev = platform_device_register_data(dev, "dmic-codec", PLATFORM_DEVID_NONE, NULL, 0);
+	if (IS_ERR(dmic_dev)) {
+		dev_err(dev, "failed to create DMIC device\n");
+		sof_pci_remove(pci);
+		return PTR_ERR(dmic_dev);
+	}
+
+	/*
+	 * We have common PCI driver probe for ACP device but we have to support I2S without SOF
+	 * for some machine. Register platform device for I2S use case without SOF if machine flag
+	 * is set to FLAG_SOF_ONLY_DMIC.
+	 */
+
+	if (flag == FLAG_SOF_ONLY_DMIC) {
```

Comment:
the bulk of this function is under this `if`. Above I suggested moving all variable declarations, only used here, inside the `if` block. Alternatively you could convert this `if` to
```
if (flag != FLAG_SOF_ONLY_DMIC)
	return 0;
```
and reduce indentation depth - if you don't plan to extend this function

Reply:
but here you only deal with the DMIC case. That seems inconsistent? what happens if the users forces the use of SOF?

Reply:
@plbossart FLAG_SOF stands for SOF usecase for DMIC + I2S endpoints while FLAG_SOF_ONLY_DMIC stands for SOF usecases with only DMIC endpoints. So DMIC is default endpoints in AMD platforms if users forces FLAG_SOF i.e "dmic-codec" platform device is registered after first case check only. When user sets for FLAG_SOF_ONLY_DMIC we will register I2S platform device as well in our pci driver.

Reply:
Sorry @ajitkupandey the meaning is different. FLAG_SOF means that SOF is selected unconditionally on a given platform. FLAG_SOF_ONLY_DMIC means that SOF is selected *only* if DMICs are detected on the platform - something we do on Intel platforms with a custom 'NHLT' table in the DSDT.

I think you are confusing the reason why SOF is chosen, and specific capabilities on a platform.

Reply:
we aren't confused but we have created flags based on use-cases and requirements on AMD platforms.

Reply:
well then call them FLAG_AMD_SOF... If you overload existing definitions with your own that's not so good.